### PR TITLE
Implement pagination

### PR DESCRIPTION
**Pull request by kevinxh**
Added pagination to Merlins and SFCC.

 **JIRA**: https://mobify.atlassian.net/browse/WEB-1353
 **Linked PRs**: https://github.com/mobify/progressive-web-sdk/pull/721

## Changes
- Added pagination component to `ProductListContent`
- Added URL query keys: `p`, `sort` and filtering option keys
Example URL: `https://www.merlinspotions.com/potions.html?p=2&sort=price`
- `sortOptions` key is added to redux store: `category` branch
- `filterOptions` key is added to redux store: `category` branch
- Four sorting options added to SFCC, see screenshot below
- Enable filtering for SFCC
- Move some functions from scaffold to SDK

## Todos
- [x] Fix filtering for pagination
- [x] Test pagination on search result page on Merlins:
~Not enough products that has more than 1 page in search result...:(~
- [x] (if applicable) analytics events instrumented to measure impact of change

## Known Bug
- [ ] On Merlins connector, cannot sort list by URL, i.e. `https://www.merlinspotions.com/potions.html?p=2&sort=price` won't sort the list, you have to manually choose sort option in the dropdown list.

## How to test-drive this PR
- Link SDK branch `pagination`
- Run `npm run dev`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- Go to potions product list page, verify the pagination is working, verify the URL is syncing correctly with the UI, verify sort is working correctly
- Do the same steps for SFCC
- On the left of app header, click `search`, search any products you like, verify pagination works in search results page

## Screenshots
screen shot 2017-06-27 at 3 44 07 pm

addtext_com_mtg1ntqzmty2mq

### CODE REVIEW

File: web/app/connectors/_sfcc-connector/categories/commands.js [hunk 1]
```diff
@@ -71,13 +89,22 @@ export const initProductListPage = (url) => (dispatch) => {
             parentId: null
         }))
     } else {
-        searchUrl = makeCategorySearchURL(categoryID)
+        searchUrl = makeCategorySearchURL(categoryID, start, sortQuery)
     }
 
     return dispatch(fetchCategoryInfo(isSearch ? null : categoryID))
```
Comment (kevinxh) [inline web/app/connectors/_sfcc-connector/categories/commands.js]: @Cdok I need your help to merge this file..:)

File: web/app/containers/product-list/partials/product-list-contents.jsx [hunk 2]
```diff
@@ -68,117 +70,165 @@ const ProductListContents = ({
...
+                                    <div className="t-product-list__filter u-flex u-margin-end-md">
+                                        <Field
+                                            idForLabel="filterButton"
+                                            label={`${numItems} Items`}
                                         >
-                                            Filter
-                                        </Button>
-                                    </Field>
+                                            <Button
+                                                className="c--tertiary u-width-full u-text-uppercase"
```
Comment (jeffkamo) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Be wary, this might conflict with Gazelle's change from #843.

Just make sure this class switches to `pw--tertiary` before merging!

**Comment by jeffkamo** [comment]
Hmm, trying to get to the PLP page but it hangs with placeholders and emits this error:

screen shot 2017-06-29 at 6 04 19 pm

I tried linking to the SDK branch with the pagination change, but I still got the same error. Am I missing step?

**Comment by gsaray** [comment]
I get these warning previewing PLP: 

image

File: web/app/utils/utils.js [hunk 3]
```diff
@@ -93,7 +124,8 @@ export const createTypedAction = (description, type, key) => createReduxAction(
 )
 
 export const buildQueryString = (query) => {
-    return query.replace(/ /g, '+')
+    return `?q=${query.replace(/ /g, '+')}`
+    // return `?q=${encodeURIComponent(query)}`
```
Comment (jeffkamo) [inline web/app/utils/utils.js]: probably can delete this comment

File: web/app/connectors/_merlins-connector/categories/parser.js [hunk 4]
```diff
@@ -36,33 +36,59 @@ const priceParser = (value) => {
...
+            if ($kind.has('.price').length > 0) {
+                const price = query.split('=')[1]
+                criteria = priceParser(price) // priceParser('10-20')
+                label = $kind.text().trim() // '$10.00 - $19.99'
+
+                // Replace `-` with `to` to prevent the replacement of non alpha-num
+                // characters from obfuscating the meaning of the query. For
+                // example, prevent `price10-20` from becoming `price1020`, because
+                // what would that mean: 10-20, 0-1020 or 1020-Infinity?
+                // 'price10to20'
```
Comment (jeffkamo) [inline web/app/connectors/_merlins-connector/categories/parser.js]: This comment is meant to be explain the result of the below variable? Can you add a line separating this and the paragraph above? Otherwise it's a bit confusing when read together.

File: web/app/connectors/_merlins-connector/categories/parser.js [hunk 5]
```diff
@@ -36,33 +36,59 @@ const priceParser = (value) => {
 }
 
 export const priceFilterParser = ($, $html) => {
```
Comment (jeffkamo) [inline web/app/connectors/_merlins-connector/categories/parser.js]: Should this be renamed to `filterParser`? It seems to handle more than just price now.

File: web/app/connectors/_merlins-connector/categories/parser.js [hunk 6]
```diff
@@ -36,33 +36,59 @@ const priceParser = (value) => {
...
+
+                // Replace `-` with `to` to prevent the replacement of non alpha-num
+                // characters from obfuscating the meaning of the query. For
+                // example, prevent `price10-20` from becoming `price1020`, because
+                // what would that mean: 10-20, 0-1020 or 1020-Infinity?
+                // 'price10to20'
+                query = query.replace(REGEX_DASH, 'to').replace(REGEX_NON_ALPHA_NUM, '')
+            } else if (query.includes('color')) { // color ruleset
+                const colorCode = query.split('=')[1]
+                criteria = {colorCode} // priceParser('10-20')}
```
Comment (jeffkamo) [inline web/app/connectors/_merlins-connector/categories/parser.js]: This comment appears out of date. Can it be adjusted to explain the `colorCode` format?

File: web/app/connectors/_sfcc-connector/categories/commands.js [hunk 7]
```diff
@@ -46,22 +64,41 @@ const fetchCategoryInfo = (id) => (dispatch) => {
...
 
 const extractCategoryId = (url) => {
-    const categoryIDMatch = /\/([^/]+)$/.exec(url)
+    const pathKeyMatch = /\/([^/]+)$/.exec(url)
+    const categoryIDMatch = pathKeyMatch ? pathKeyMatch[1].match(/([^?]*)|([^?]*)\?.*/) : ''
     return categoryIDMatch ? categoryIDMatch[1] : ''
 }
 
+const extractPageNumber = (url) => {
+    const pageMatch = url.match(/p=(.*)/)
```
Comment (jeffkamo) [inline web/app/connectors/_sfcc-connector/categories/commands.js]: I suppose we are fairly sure there won't be any other queries after `p=` here?

File: web/app/connectors/_sfcc-connector/parsers.js [hunk 8]
```diff
@@ -159,6 +159,15 @@ export const parseProductListData = (products) => {
     return productListData
 }
 
+export const parseSortedProductKeys = (products) => {
+    const sortedProductKeys = []
+
+    products.forEach((productHit) => {
```
Comment (jeffkamo) [inline web/app/connectors/_sfcc-connector/parsers.js]: Minor: this is more out of curiosity, but is there a reason you used `forEach` instead of map in this case?

File: web/app/containers/product-list/partials/product-list-contents.jsx [hunk 9]
```diff
@@ -68,118 +70,164 @@ const ProductListContents = ({
...
+                                            >
+                                                <select
+                                                    className="u-color-neutral-60"
+                                                    value={selectedSortOption}
+                                                    onChange={(e) => { updateURL({sort: e.target.value}) }}
+                                                    onBlur={(e) => { updateURL({sort: e.target.value}) }}
+                                                >
+                                                    {
+                                                        sortOptions.map((choice, index) => <option key={index} value={choice.id}>{choice.label}</option>)
+                                                    }
```
Comment (jeffkamo) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Minor: the format we've used for this sort of thing is like this:

```jsx
{sortOptions.map((a, b) =>
    <option {...} />
)}
```

This might help keep your line length down to a bit more manageable size.

File: web/app/containers/product-list/partials/product-list-contents.jsx [hunk 10]
```diff
@@ -68,118 +70,164 @@ const ProductListContents = ({
...
+        browserHistory.push({pathname, query})
+    }
+
+    const clearFiltersURL = () => {
+        clearFilters()
+        const query = Object.assign({}, location.query)
+        delete query.price
+        delete query.color
+        browserHistory.push({pathname, query})
+    }
```
Comment (jeffkamo) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Hmm, I'd like to ask for an Engineer to check in on these. I'm wondering: is okay if we are calling `browserHistory` directly from the UI like this, or should it be called from an Action?

File: web/app/containers/product-list/selectors.js [hunk 11]
```diff
@@ -21,30 +20,49 @@ export const getCurrentProductList = createGetSelector(
...
+
+/* filtering is done on page
 export const getFilteredProductListProducts = createSelector(
     getCategoryProducts,
     getActiveFilters,
     (products, filters) => {
         return filters.size > 0 ? products.filter(byFilters(filters.toJS())) : products
     }
-)
+)*/
```
Comment (jeffkamo) [inline web/app/containers/product-list/selectors.js]: This can probably be deleted if it's not being kept.

File: web/app/containers/product-list/selectors.js [hunk 12]
```diff
@@ -21,30 +20,49 @@ export const getCurrentProductList = createGetSelector(
...
-    getFilters,
+    getCategoryFilterOptions,
     (filters) => (
         filters.reduce((activeFilters, filter) => activeFilters.concat(
             filter.get('kinds').filter((kind) => kind.get('active'))
         ), Immutable.List())
     )
 )
+
+/* filtering is done on page
```
Comment (jeffkamo) [inline web/app/containers/product-list/selectors.js]: Does this mean we're manipulating data in the UI?

File: web/app/containers/product-list/selectors.js [hunk 13]
```diff
@@ -21,30 +20,49 @@ export const getCurrentProductList = createGetSelector(
...
     getActiveFilters,
     (products, filters) => {
         return filters.size > 0 ? products.filter(byFilters(filters.toJS())) : products
     }
-)
+)*/
 
 export const getFilteredAndSortedListProducts = createSelector(
-    getFilteredProductListProducts,
+    getCategoryProducts,
```
Comment (jeffkamo) [inline web/app/containers/product-list/selectors.js]: If this is no longer filtering, the name of the function could probably be updated

File: web/app/connectors/_sfcc-connector/categories/commands.js (hunk 7 above)
Comment (wizardlyhel) [inline web/app/connectors/_sfcc-connector/categories/commands.js]: We cannot assume that. The moment the order of the query param changes, this piece of code will break.

Such regex match should account for:
`https://www.merlinspotions.com/search?p=3`
`https://www.merlinspotions.com/search?c=5&p=3`
`https://www.merlinspotions.com/search?c=5&p=3&d=test`
`https://www.merlinspotions.com/search?p=3&c=5&d=test`
`https://www.merlinspotions.com/search?c=5&d=test&p=3`
`https://www.merlinspotions.com/search?c=5&d=test&p=3#test`
^ since the string is a url and not just the query param, we need to account for hash as well

File: web/app/connectors/_merlins-connector/categories/commands.js [hunk 14]
```diff
@@ -2,33 +2,55 @@
...
 } from 'progressive-web-sdk/dist/integration-manager/categories/results'
 import {receiveProductListProductData} from 'progressive-web-sdk/dist/integration-manager/products/results'
-import categoryProductsParser, {parseCategoryTitle, parseCategoryId, priceFilterParser} from './parser'
+import {changeFilterTo} from '../../../store/categories/actions'
 import {productListParser} from '../products/parsers'
 import {getTextFrom} from '../../../utils/parser-utils'
 import {fetchPageData} from '../app/commands'
 
+const extractFilter = (url) => {
+    const filter = url.match(/(price=[^&|#]*)/) || url.match(/(color=[^&|#]*)/)
```
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/categories/commands.js]: Yes, Merlins can only have one filter at a time before this PR, we can expand this to support multiple filters but I think it is out of scope of the ticket...

File: web/app/containers/product-list/partials/product-list-contents.jsx (hunk 10 above)
Comment (kevinxh) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: @jeffkamo Yea, I did feel like that's a bit hacky, but not sure what's the proper way of handling URL changes.

@jeremywiebe Any idea about how should I approach this?

File: web/app/containers/product-list/partials/product-list-contents.jsx [hunk 15]
```diff
@@ -68,118 +70,164 @@ const ProductListContents = ({
...
+                                            <Field
+                                                className="pw--has-select"
+                                                idForLabel="sort"
+                                                label="Sort by"
+                                            >
+                                                <select
+                                                    className="u-color-neutral-60"
+                                                    value={selectedSortOption}
+                                                    onChange={(e) => { updateURL({sort: e.target.value}) }}
+                                                    onBlur={(e) => { updateURL({sort: e.target.value}) }}
```
Comment (wizardlyhel) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: This requires `data-analytics-name={UI_NAME.sortBy}` for analytics

File: web/app/connectors/_sfcc-connector/parsers.js (hunk 8 above)
Comment (kevinxh) [inline web/app/connectors/_sfcc-connector/parsers.js]: This is because I want to preserve the order of the list, since the sorting is done on the server side. `map` returns an object that is ordered by it's keys, in this case, it is order by `produceId`.

File: web/app/utils/utils.js (hunk 3 above)
Comment (kevinxh) [inline web/app/utils/utils.js]: Got it.

File: web/app/connectors/_merlins-connector/categories/parser.js (hunk 4 above)
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/categories/parser.js]: Fixed.

File: web/app/connectors/_merlins-connector/categories/parser.js (hunk 5 above)
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/categories/parser.js]: Fixed.

File: web/app/connectors/_merlins-connector/categories/parser.js (hunk 6 above)
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/categories/parser.js]: Fixed.

File: web/app/containers/product-list/partials/product-list-contents.jsx [hunk 16]
```diff
@@ -68,118 +70,164 @@ const ProductListContents = ({
...
+                    <NoResultsList routeName={routeName} />
+                }
+                {page && pageCount > 1 &&
+                    <Pagination
+                        className="u-margin-top-lg"
+                        onChange={(newPage) => updateURL({p: newPage})}
+                        currentPage={page ? page : 1}
+                        pageCount={pageCount}
+                        showCurrentPageMessage={true}
+                        showPageButtons={false}
```
Comment (wizardlyhel) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Pagination analytics event needs to be instrumented at the SDK level:

previous button -
```jsx
data-analytics-name={UI_NAME.pagination} data-analytics-content={UI_NAME.previous}
```

next button -
```jsx
data-analytics-name={UI_NAME.pagination} data-analytics-content={UI_NAME.next}
```

numbered button -
```jsx
data-analytics-name={UI_NAME.pagination} data-analytics-content=`page ${pageNumber}`
```

File: web/app/containers/product-list/partials/product-list-contents.jsx (hunk 9 above)
Comment (kevinxh) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: 👍

File: web/app/containers/product-list/selectors.js (hunk 13 above)
Comment (kevinxh) [inline web/app/containers/product-list/selectors.js]: Good catch, fixed.

File: web/app/connectors/_merlins-connector/categories/commands.js [hunk 17]
```diff
@@ -2,33 +2,55 @@
...
             const searchTermMatch = title.match(/'(.*)'/)
+            const sortOptions = parseSortOptions($, $response)
+            const isFilter = parseCurrentFilter($, $response)
+            if (sortOptions.length > 0) {
+                dispatch(receiveCategorySortOptions(sortOptions, pathKeyWithoutQuery))
+            }
+            if (!isFilter) {
+                dispatch(receiveCategoryFilterOptions(priceFilterParser($, $response), pathKeyWithoutQuery))
+            } else {
+                dispatch(changeFilterTo(filter))
```
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/categories/commands.js]: On Merlins Desktop site, if there is a selected filter, then all filter options disappear unless I clear the filter. That said, if a user goes to `http://www.merlinspotions.com/?price=10-20` on mobile, then something will break since no filter options have been received...

Any better ideas?

File: web/app/connectors/_merlins-connector/config.js [hunk 18]
```diff
@@ -9,9 +9,11 @@ export const CART_URL = '/checkout/cart/'
 export const PAYMENT_URL = '/checkout/payment/'
 export const CREATE_ACCOUNT_POST_URL = '/customer/account/createpost/'
 export const LOGIN_POST_URL = '/customer/account/loginPost/'
+export const SEARCH_URL = '/catalogsearch/result/'
+export const SEARCH_SUGGESTION_URL = '/search/ajax/suggest/'
```
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/config.js]: fixed.

File: web/app/connectors/_merlins-connector/categories/parser.js [hunk 19]
```diff
@@ -79,4 +105,22 @@ const categoryProductsParser = ($, $html) => {
...
+    $select.children().each((index, option) => {
+        sortOptions.push({
+            id: option.value,
+            label: option.text
+        })
+    })
+    return sortOptions
+}
+
+export const parseCurrentFilter = ($, $html) => {
```
Comment (kevinxh) [inline web/app/connectors/_merlins-connector/categories/parser.js]: fixed.

File: web/app/connectors/_sfcc-connector/categories/commands.js (hunk 7 above)
Comment (kevinxh) [inline web/app/connectors/_sfcc-connector/categories/commands.js]: Thanks @wizardlyhel, fixed.

Comment (kevinxh) [inline web/app/connectors/_sfcc-connector/categories/commands.js]: fixed.

File: web/app/connectors/_sfcc-connector/categories/commands.js [hunk 20]
```diff
@@ -46,22 +64,41 @@ const fetchCategoryInfo = (id) => (dispatch) => {
...
+    return pageMatch ? pageMatch[1] : '1'
+}
+
+const extractSortOption = (url) => {
+    const sortOption = url.match(/sort=(.*)/)
```
Comment (kevinxh) [inline web/app/connectors/_sfcc-connector/categories/commands.js]: fixed.

File: web/app/utils/utils.js [hunk 21]
```diff
@@ -30,6 +30,37 @@ export const getPath = ({pathname, search}) => pathname + search
...
+ */
+export const makeQueryString = (params) => {
+    if (typeof params === 'undefined' || typeof params !== 'object') {
+        return ''
+    }
+
+    let query = '?'
+    let index = 0
+
+    for (const i in params) {
```
Comment (kevinxh) [inline web/app/utils/utils.js]: Good idea, addressed.

File: web/app/containers/product-list/partials/product-list-contents.jsx (hunk 16 above)
Comment (kevinxh) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Thanks for the clear instruction!

I also added you as a reviewer to the SDK PR.

File: web/app/containers/product-list/partials/product-list-contents.jsx (hunk 15 above)
Comment (kevinxh) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Fixed.

**Comment by wizardlyhel** [comment]
👍 pagination analytics looks good

File: web/app/containers/product-list/partials/product-list-contents.jsx (hunk 10 above)
Comment (jeremywiebe) [inline web/app/containers/product-list/partials/product-list-contents.jsx]: Well, I'd like to keep `browserHistory` out of the IM and connectors (although a few instances exist). The only reason I'd see to extract these things into actions is so that they can be tested more easily.  That said, it wouldn't be hard to mock `browserHistory` and test this component directly.  

TLDR; I'm ok with this.

File: web/app/connectors/_merlins-connector/categories/commands.js (hunk 17 above)
Comment (jeremywiebe) [inline web/app/connectors/_merlins-connector/categories/commands.js]: I can't think of a better way either. Since this is within the Merlin's connector let's just leave it.

**Comment by jeremywiebe** [comment]
Can we fix up how the item count looks?  I had to look at it twice to understand it was saying there was 1 matching item for "Green" potions: 

image

**Comment by jeremywiebe** [comment]
We should be able to add filters (called "refinements" in SFCC) also. If you look at the response to the `product_search` call, there's a `refinements` key that defines the types of filtering you can do on the given base product search. From there it also includes the discrete filter values.

image

**Comment by mitchquik** [comment]
Optimal UX would be to display the number of items in the list below the list of filters rather than next to the filters themselves:
screen shot 2017-07-06 at 10 46 49 am

**Comment by jeremywiebe** [comment]
@mitchquik With regards to the location of the count, it's a bit complicated because without any filters applied there is no "current filter" area and so it seems like the count is in the right place in that situation. 

image
